enhancements/20240805-encrypted-block-volume.md

```diff
@@ -0,0 +1,440 @@
+  We could verify this by checking `/dev/longhorn` in the host like following screenshot after starting VM.  
+  If it's encrypted, it shows LUKS header information. Otherwise, it shows invalid LUKS device.
+  ![](./20240805-encrypted-block-volume/06.png)
+
+  Then, you could use wrong password to `lukeOpen` it.  
+  ![](./20240805-encrypted-block-volume/07.png)
+  
+
+### Limitations
+
+- Don't support that export to image from encrypted boot and data volumes. 
+- Don't support that download encrypted image and upload it to reuse.
+- Don't support that data volume restoring from encrypted to unencrypted or unencrypted to encrypted one.
+
+
+### Upgrade strategy
+
+None
+
+## Notes
+
+In some discussions with @bk201 @Vicente-Cheng @WebberHuang1118 , we considered to follow CSI interface for encryption. For example, Harvester only requires that CSI driver provides Harvester  with an encrypted volume. Harvester would attach it to trigger the CSI driver's decryption. Next, Harvester would copy image data to this volume.
+
+However, this approach could lead to performance issue. With this way, Harvester needs to copy image data to encrypted volume each time when starting VMs, but underlying backing image does not. This is not efficient. Although we could use `snapshot` to avoid this, our `virtualmachineimages.harvesterhci.io` is tightly coupled with Longhorn. This would result in more steps for end-users to create encrypted virtual machine image.
+
+Therefore, at this moment, we've decided to follow Longhorn's method for encryption/decryption until we find a better solution.
+
+## FAQ
+
+### Q1: If we use k8s secret to encrypt volume, how to prevent different users from reading the secret?
+
+Depend on environment. 
+
+If we use standalone Harvester, there is only one user. So, we can't avoid this.
+
+If we use Harvester with Rancher Integration, we could assign different users to different projects and namespaces. By this, we could prevent different users from reading the other secret. Please check [Multi-Tenancy with Rancher Integration](https://docs.harvesterhci.io/v1.3/rancher/virtualization-management/#multi-tenancy-example).
+
+### Q2: Can we prevent users from creating Storage Class?
+
+Yes. Please follow answer of Q1 to use Rancher Integration to create different projects and namespaces. Storage class is available to everyone, normal users can still read it. But, they can't create it.
+
+The only way for normal users to create Storage Class is to request admin to create it.
+
+### Q3: How do we ***really*** know this volume is encrypted or not?
+
+For end-user, they never know. The CSI driver automatically decrypts the volume when volumes are attached to VMs.
+
+But, for admin which can log in to host, they could check `/dev/longhorn/pvc-xxxx` by command `cryptsetup luksDump {pvc_name}` after VMs start. Please check screenshots of [Test Plan](#test-plan).
```

Comment:
Do we have any field in image & volume APIs response indicates the image & volume is encrypted or not?  If have, we can show a lock icon before image/volume name or add a new volumn named encrypted.

See longhorn screenshot
<img width="1490" alt="Screenshot 2024-09-05 at 11 45 35 AM" src="https://github.com/user-attachments/assets/27d06628-e472-4299-a70a-6f48f2884ee1">

Reply:
About image, you could use securityParameters as condition, if `spec.securityParameters.cryptoOperation` is `encrypt`, then it's a encrypted image.

About volume, as I know, `v1/harvester/persistentvolumes` and `v1/harvester/longhorn.io.volumes` both contain a fields call `encrypted`. One is under `spec.csi.volumeAttributes.encrypted` of `v1/harvester/persistentvolumes`, another one is under `spec.encrypted` of `v1/harvester/longhorn.io.volumes`.

Reply:
I agree with adding a lock icon as well.